File: Loenn/entities/spinyPlant.lua
local drawableSpriteStruct = require("structs.drawable_sprite")
local drawing = require("utils.drawing")
local utils = require("utils")
local enums = require("consts.celeste_enums")

local function getTexture(entity)
    return entity.texture and entity.texture ~= "default" and entity.texture or "wood"
end

local plant = {}

plant.name = "JungleHelper/SpinyPlant"
plant.depth = -9500
plant.canResize = {false, true}
plant.minimumSize = { 0, 16 }

local colors = { "Blue", "Pink", "Yellow", "Orange" }

plant.placements = {}

for _, color in ipairs(colors) do
    table.insert(plant.placements, {
        name = color,
        data = {
            height = 16,
            color = color,
            sprite = ""
        },
    })
end

plant.fieldInformation = {
    color = {
        options = colors,
        editable = false
    }
}

function plant.sprite(room, entity)
    local textureRaw = getTexture(entity)
    local texture = "objects/plant/" .. textureRaw

    local x, y = entity.x or 0, entity.y or 0
    local height = entity.height or 8

    local startX, startY = math.floor(x / 8) + 1, math.floor(y / 8) + 1
    local stopY = startY + math.floor(height / 8) - 1
    local len = stopY - startY

    local sprites = {}

    for i = 0, len, 2 do
        local topOpen = room.tilesFg.matrix:get(startX + 1, startY - 1, "0") == "0"
        local bottomOpen = room.tilesFg.matrix:get(startX + 1, stopY + 1, "0") == "0"

        local section = ""
        if len == 1 then
            if topOpen then
                if bottomOpen then
                    section = "Solo"
                else
                    section = "Top"
                end
            else
                if bottomOpen then
                    section = "Bottom"
                else
                    section = "Mid"
                end
            end
        else
            if i == len then
                i = len - 1
            end

            section = "Mid"
            if i == 0 and topOpen then
                section = "Top"
            elseif i == len - 1 and bottomOpen then
                section = "Bottom"
            end
        end

        -- We cannot pass the entity to drawableSprite.fromTexture, because otherwise its "color" option is going to be interpreted as tinting. :a:
        local sprite = drawableSpriteStruct.fromTexture("JungleHelper/SpinyPlant/Spiny" .. entity.color .. section, {x = entity.x, y = entity.y})

        sprite:setJustification(0, 0)
        sprite:addPosition(section == "Solo" and 4 or 0, i * 8)
        table.insert(sprites, sprite)
    end

    return sprites
end

function plant.selection(room, entity)
    return utils.rectangle(entity.x, entity.y, 24, entity.height)
end

return plant

File: Ahorn/entities/spinyPlant.jl
module JungleHelperSpinyPlant

using ..Ahorn, Maple

@mapdef Entity "JungleHelper/SpinyPlant" SpinyPlant(x::Integer, y::Integer, height::Integer=16, color::String="Blue", sprite::String="")

colors = ["Blue", "Pink", "Yellow", "Orange"]

const placements = Ahorn.PlacementDict(
    "Spiny Plant ($(color)) (Jungle Helper)" => Ahorn.EntityPlacement(
        SpinyPlant,
        "rectangle",
        Dict{String, Any}(
            "color" => color
        )
    ) for color in colors
)

Ahorn.editingOptions(entity::SpinyPlant) = Dict{String, Any}(
    "color" => colors
)

Ahorn.minimumSize(entity::SpinyPlant) = 0, 16
Ahorn.resizable(entity::SpinyPlant) = false, true

function Ahorn.selection(entity::SpinyPlant)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 16))

    return Ahorn.Rectangle(x, y, 24, height)
end

sections = ["Top", "Mid", "Bottom"]

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::SpinyPlant, room::Maple.Room)
    color = get(entity.data, "color", "Blue")

    # Values need to be system specific integer
    x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    height = Int(get(entity.data, "height", 16))
    left = get(entity.data, "left", true)

    startX = div(x, 8) + 1
    startY = div(y, 8) + 1
    stopY = startY + div(height, 8) - 1

    len = stopY - startY

    if len == 1
        topOpen = get(room.fgTiles.data, (startY - 1, startX + 1), false) == '0'
        bottomOpen = get(room.fgTiles.data, (stopY + 1, startX + 1), false) == '0'

        if topOpen
            if bottomOpen
                section = "Solo"
            else
                section = "Top"
            end
        else
            if bottomOpen
                section = "Bottom"
            else
                section = "Mid"
            end
        end

        Ahorn.drawImage(ctx, "JungleHelper/SpinyPlant/Spiny$(color)$(section)", section == "Solo" ? 4 : 0, 0)
    else
        for i in 0:2:len
            if i == len
                i = len - 1
            end

            section = "Mid"
            if i == 0
                if get(room.fgTiles.data, (startY - 1, startX + 1), false) == '0'
                    section = "Top"
                end
            elseif i == len - 1
                if get(room.fgTiles.data, (stopY + 1, startX + 1), false) == '0'
                    section = "Bottom"
                end
            end

            Ahorn.drawImage(ctx, "JungleHelper/SpinyPlant/Spiny$(color)$(section)", 0, 8 * i)
        end
    end
end

end
